README.md
[![Travis Status][trav_img]][trav_site]
[![Appveyor Status][av_img]][av_site]
[![Coverage Status][cov_img]][cov_site]
[![Maintenance Status][maintenance-image]](#maintenance-status)


Builder
=======

Builder takes your `npm` tasks and makes them composable, controllable from
a single point, and flexible.

`npm` is fantastic for controlling tasks (via `scripts`) and general project
workflows. But a project-specific `package.json` simply doesn't scale when
you're managing many (say 5-50) very similar repositories.

_Enter Builder._ Builder is "almost" `npm`, but provides for off-the-shelf
"archetypes" to provide central sets of `package.json` `scripts` tasks, and
`dependencies` and `devDependencies` for those tasks. The rest of this page will
dive into the details and machinations of the tool, but first here are a few of
the rough goals and motivations behind the project.

* **Single Point of Control**: A way to define a specific set of tasks /
  configs / etc. for one "type" of project. For example, we have an
  ever-expanding set of related repos for our
  [Victory](https://github.com/FormidableLabs/?utf8=%E2%9C%93&query=victory)
  project which all share a nearly-identical dev / prod / build workflow.
* **Flexibility**: There are a number of meta tools for controlling JavaScript
  workflows / development lifecycles. However, most are of the "buy the farm"
  nature. This works great when everything is within the workflow but falls
  apart once you want to be "just slightly" different. Builder solves this by
  allowing fine grain task overriding by name, where the larger composed tasks
  still stay the same and allow a specific repo's deviation from "completely off
  the shelf" to be painless.
* **You Can Give Up**: One of the main goals of builder is to remain very
  close to a basic `npm` workflow. So much so, that we include a section in this
  guide on how to abandon the use of Builder in a project and revert everything
  from archetypes back to vanilla `npm` `package.json` `scripts`, `dependencies`
  and `devDependencies`.
* **A Few "Nice to Haves" Over `npm run <task>`**: Setting aside archetypes and
  multi-project management, `builder` provides cross-OS compatible helpers for
  common task running scenarios like concurrent execution (`concurrent`) and
  spawning the _same_ tasks in parallel with different environment variables
  (`env`). It also provides useful controls for task retries, buffered output,
  setup tasks, etc.

**Contents**:

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->


- [Overview](#overview)
  - [Usage](#usage)
    - [Global Install](#global-install)
    - [Local Install](#local-install)
      - [PATH Augmentation](#path-augmentation)
      - [Full Path Invocation](#full-path-invocation)
    - [Configuration](#configuration)
    - [Builder Actions](#builder-actions)
      - [builder run](#builder-run)
      - [builder concurrent](#builder-concurrent)
      - [builder envs](#builder-envs)
    - [Setup Task](#setup-task)
    - [Task Lifecycle](#task-lifecycle)
      - [The Basics](#the-basics)
      - [Other Builder Actions](#other-builder-actions)
      - [Builder Flags During Pre and Post](#builder-flags-during-pre-and-post)
      - [Task Prefix Complexities](#task-prefix-complexities)
    - [Custom Flags](#custom-flags)
    - [Expanding the Archetype Path](#expanding-the-archetype-path)
- [Tasks](#tasks)
- [npm Config](#npm-config)
  - [`npm` Config Overview](#npm-config-overview)
  - [Builder Configs](#builder-configs)
  - [Config Notes](#config-notes)
    - [Tip - Use String Values](#tip---use-string-values)
    - [npmrc Configuration](#npmrc-configuration)
    - [Command Line Environment Variables](#command-line-environment-variables)
- [Archetypes](#archetypes)
  - [Task Resolution](#task-resolution)
  - [Special Archetype Tasks](#special-archetype-tasks)
  - [Creating an Archetype](#creating-an-archetype)
    - [Initializing a Project](#initializing-a-project)
    - [Managing the `dev` Archetype](#managing-the-dev-archetype)
    - [Node Require Resolution and Module Pattern](#node-require-resolution-and-module-pattern)
      - [The Module Pattern](#the-module-pattern)
        - [ES.next Imports and The Module Pattern](#esnext-imports-and-the-module-pattern)
        - [Webpack and Module Pattern](#webpack-and-module-pattern)
    - [Application vs. Archetype Dependencies](#application-vs-archetype-dependencies)
    - [Moving `dependencies` and `scripts` to a New Archetype](#moving-dependencies-and-scripts-to-a-new-archetype)
      - [Moving `dependencies` and `devDependencies` from an Existing `package.json`](#moving-dependencies-and-devdependencies-from-an-existing-packagejson)
      - [Moving `scripts` and Config Files](#moving-scripts-and-config-files)
      - [Updating Path and Module References in Config Files](#updating-path-and-module-references-in-config-files)
    - [Example `builder` Archetype Project Structure](#example-builder-archetype-project-structure)
- [Tips, Tricks, & Notes](#tips-tricks--notes)
  - [PATH, NODE_PATH Resolution](#path-node_path-resolution)
  - [Environment Variables](#environment-variables)
  - [Alternative to `npm link`](#alternative-to-npm-link)
  - [Project Root](#project-root)
  - [Avoid npm Lifecycle Commands](#avoid-npm-lifecycle-commands)
  - [Other Process Execution](#other-process-execution)
  - [I Give Up. How Do I Abandon Builder?](#i-give-up-how-do-i-abandon-builder)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Overview

At a high level `builder` is a tool for consuming `package.json` `scripts`
commands, providing sensible / flexible defaults, and supporting various scenarios
("archetypes") for your common use cases across multiple projects.

Builder is not opinionated, although archetypes _are_ and typically dictate
file structure, standard configurations, and dev workflows. Builder supports
this in an agnostic way, providing essentially the following:

* `NODE_PATH`, `PATH` enhancements and module patterns to run, build, import
  from archetypes so task dependencies and configurations don't have to be
  installed directly in a root project.
* A task runner capable of single tasks (`run`) or multiple concurrent tasks
  (`concurrent`).
* An intelligent merging of `package.json` `scripts` tasks.

... and that's about it!

### Usage

To start using builder, install and save `builder` and any archetypes you
intend to use. We'll use the [builder-react-component][] archetype as an
example.

**Note**: Most archetypes have an `ARCHETYPE` package and parallel
`ARCHETYPE-dev` npm package. The `ARCHETYPE` package contains _almost_
everything needed for the archetype (prod dependencies, scripts, etc.) except
for the `devDependencies` which the latter `ARCHETYPE-dev` package is solely
responsible for bringing in.

#### Global Install

For ease of use, one option is to globally install `builder` and locally install
archetypes:

```sh
$ npm install -g builder
$ npm install --save builder-react-component
$ npm install --save-dev builder-react-component-dev
```

Like a global install of _any_ Node.js meta / task runner tool (e.g., `eslint`,
`mocha`, `gulp`, `grunt`) doing a global install is painful because:

* You are tied to _just one_ version of the tool for all projects.
* You must also globally install the tool in CI, on servers, etc.

... so instead, we **strongly recommend** a local install described in the
next section!

To help you keep up with project-specific builder requirements, a globally-installed
`builder` will detect if a locally-installed version of `builder` is
available and switch to that instead:

```sh
$ /GLOBAL/PATH/TO/builder
[builder:local-detect] Switched to local builder at: ./node_modules/builder/bin/builder-core.js

... now using local builder! ...
```

#### Local Install

To avoid tying yourself to a single, global version of `builder`, the option
that we endorse is locally installing both `builder` and archetypes:

```sh
$ npm install --save builder
$ npm install --save builder-react-component
$ npm install --save-dev builder-react-component-dev
```

However, to call `builder` from the command line you will either need to either
augment `PATH` or call the long form of the command:

##### PATH Augmentation

Our recommended approach is to augment your `PATH` variable with a shell
configuration as follows:

**Mac / Linux**

```sh
# Safer version, but if you _have_ global installs, those come first.
export PATH="${PATH}:./node_modules/.bin"

# (OR) Less safe, but guarantees local node modules come first.
export PATH="./node_modules/.bin:${PATH}"

# Check results with:
echo $PATH
```

To make these changes **permanent**, add the `export` command to your `.bashrc`
or analogous shell configuration file.

**Windows**

```sh
# Safer version, but if you _have_ global installs, those come first.
set PATH=%PATH%;node_modules\.bin

# (OR) Less safe, but guarantees local node modules come first.
set PATH=node_modules\.bin;%PATH%

# Check results with:
echo %PATH%
```

To make these changes **permanent**, please see this multi-OS article on
changing the `PATH` variable: https://www.java.com/en/download/help/path.xml
(the article is targeted for a Java executable, but it's analogous to our
situation). You'll want to paste in `;node_modules\.bin` at the end _or_
`node_modules\.bin;` at the beginning of the PATH field in the gui. If there
is no existing `PATH` then add a user entry with `node_modules\.bin` as a value.
(It is unlikely to be empty because an `npm` installation on Windows sets the
user `PATH` analogously.)

##### Full Path Invocation

Or you can run the complete path to the builder script with:

**Mac / Linux**

```sh
node_modules/.bin/builder <action> <task>
```

**Windows**

```sh
node_modules\.bin\builder <action> <task>
```

#### Configuration

After `builder` is available, you can edit `.builderrc` like:

```yaml
---
archetypes:
  - builder-react-component
```

to bind archetypes.

... and from here you are set for `builder`-controlled meta goodness!

#### Builder Actions

Display general or command-specific help (which shows available specific flags).

```sh
$ builder [-h|--help|help]
$ builder help <action>
$ builder help <archetype>
```

Run `builder help <action>` for all available options. Version information is
available with:

```sh
$ builder [-v|--version]
```

Let's dive a little deeper into the main builder actions:

##### builder run

Run a single task from `script`. Analogous to `npm run <task>`

```sh
$ builder run <task>
```

Flags:

* `--tries`: Number of times to attempt a task (default: `1`)
* `--setup`: Single task to run for the entirety of `<action>`
* `--quiet`: Silence logging
* `--log-level`: Level to log at (`debug`, `info`, `warn`, `error`, `none`) (default: `error`)
* `--env`: JSON object of keys to add to environment.
* `--env-path`: JSON file path of keys to add to environment.
* `--expand-archetype`: Expand `node_modules/<archetype>` with full path (default: `false`)
* `--builderrc`: Path to builder config file (default: `.builderrc`)

##### builder concurrent

Run multiple tasks from `script` concurrently. Roughly analogous to
`npm run <task1> & npm run <task2> & npm run <task3>`, but kills all processes on
first non-zero exit (which makes it suitable for test tasks), unless `--no-bail`
is provided.

```sh
$ builder concurrent <task1> <task2> <task3>
```

Flags:

* `--tries`: Number of times to attempt a task (default: `1`)
* `--setup`: Single task to run for the entirety of `<action>`.
    * **Note**: The `--setup` task is run at the start of the first main task
      to actually run. This may _not_ be the first specified task however,
      as `pre` tasks could end up with main tasks starting out of order.

HISTORY.md
History
=======

## 5.0.0

**BREAKING**:

* Restrict to `node` version `8+`.
* Some modest ESnext refactoring.

## 4.0.0

**BREAKING**:

* Restrict to `node` version `4+`.
* Default to `--log-level=error` intead of `info`.
* Add `pre|post` lifecycle commands.
  [#68](https://github.com/FormidableLabs/builder/issues/68)
* Error out on invalid or conflicting command line flags passed
  to `builder`.

Other enhancements:

* Add `--log-level=debug`. Switch some existing logs over.
* Add more explicit behavior and tests for `--setup` flag.

## 3.2.3

* Account for `PATH` vs. `Path` env var differences in MinGW vs. cmd on
  Windows.
  [#142](https://github.com/FormidableLabs/builder/issues/142)

## 3.2.2

* Correctly infer if a builder passthrough task on Windows. ( [@GEMI][] )
  [#138](https://github.com/FormidableLabs/builder/issues/138)

## 3.2.1

* Update module pattern documentation.

## 3.2.0

* Add in `NODE_PATH` and `PATH` from _actual resolved paths_ of prod and dev
  archetypes instead of guessed paths. This will better support `yarn`, which
  flattens `node_modules/.bin` in different ways than real `npm`. It is also
  likely more correct than before.
  [#134](https://github.com/FormidableLabs/builder/issues/134)
* Add extra higher level directory check when `LOCAL_DEV=true` and
  `--expand-archetype` specified.
* Make config loading failure a simple `log.info` instead of `log.warn`.

## 3.1.0

* Add `--env` environment variable flag.
  [#45](https://github.com/FormidableLabs/builder/issues/45)

## 3.0.0

**BREAKING**:

* Remove `--unlimited-buffer` command line option. It is unneeded with switch to
  `spawn`.

Other:

* Default log level to `none` for `help` and `version` commands.
  [#127](https://github.com/FormidableLabs/builder/issues/127)
* Switch from `exec` to `spawn` child process spawning for better stdio/stderr
  propagation. ( [@exogen][] )
  [#20](https://github.com/FormidableLabs/builder/issues/20)
* Add auto-TOC to README.md.
  [#101](https://github.com/FormidableLabs/builder/issues/101)
* Use a "safe exit" pattern to avoid immediately exiting the `builder` process
  when there is potentially unflushed output.
  [#124](https://github.com/FormidableLabs/builder/issues/124)

## 2.10.1

* Fix error from config in archetype but not root package.json.
  [#113](https://github.com/FormidableLabs/builder/issues/113)

## 2.10.0

* Add `--unlimited-buffer` shell output option.

## 2.9.1

* Refactor lodash calls to be tolerant of v3 or v4 as a temporary bandaid to
  a solution for the real underlying issue of:
  [#99](https://github.com/FormidableLabs/builder/issues/99)

## 2.9.0

* Add configurable log levels via `--log-level=LEVEL` and `--quiet`.
  [#6](https://github.com/FormidableLabs/builder/issues/6)

## 2.8.0

* Revises `PATH`, `NODE_PATH` ordering to place archetype first, then root
  project.
* Add `--expand-archetype` flag to expand `node_modules/<archetype` tokens in
  task strings.
  [builder-victory-component#23](https://github.com/FormidableLabs/builder-victory-component/issues/23)

## 2.7.1

* Propagate `--` flags via environment instead of command line mutation.
  [#92](https://github.com/FormidableLabs/builder/issues/92)

## 2.7.0

* Add support for `package.json:config` analogous to `npm`.
  [#89](https://github.com/FormidableLabs/builder/issues/89)

## 2.6.0

* Fix bug wherein `builder --version` displayed help instead of version.
* Add significantly enhanced test coverage.

## 2.5.0

* Add `builder -v|--version`.
* Add `builder -h|--help` shortcuts.
* Make `builder` (with no args) display help.
  [#61](https://github.com/FormidableLabs/builder/issues/61)

## 2.4.0

* Add `builder <action> <task> [-- <args>...]` support.
  [builder-react-component#27](https://github.com/FormidableLabs/builder-react-component/issues/27)

## 2.3.3

* Harmonize log messages with standardized command + environment strings.
* Add `builder <action> --[no-]bail` flag to control failing vs. non-failing
  concurrent tasks. ( [@exogen][] )
  [#64](https://github.com/FormidableLabs/builder/issues/64)

## 2.3.2

* Fix `stdout maxBuffer exceeded` error by bumping `exec` buffer to 32mb.
  [#62](https://github.com/FormidableLabs/builder/issues/62)

## 2.3.1

* Fix `builder run envs` on Node `v4.x` with respect to
  `_.isPlainObject(process.env)` bug. ( [@exogen][] )
  [#63](https://github.com/FormidableLabs/builder/issues/63)

## 2.3.0

* Switch to `tree-kill` to more emphatically kill off children process spawned
  within a single builder run.
* Add `builder <action> --setup` flag to run a task before and during an action.
  [#51](https://github.com/FormidableLabs/builder/issues/51)

## 2.2.2

* Fix help to work with `builder help <action|archetype(s)>`.

## 2.2.1

* Minor documentation and usage updates.

## 2.2.0

* Global `builder` script detects if local install available and switches to it.
  [#10](https://github.com/FormidableLabs/builder/issues/10)
* Add `builder envs` action for concurrent runs based on environment variables.
  [#29](https://github.com/FormidableLabs/builder/issues/29)
* Add `builder envs|concurrent --buffer` flag to buffer and display stderr and
  stdout at the _end_ of a task run for easier reading of concurrent output.

## 2.1.4

* Fix `PATH` / `NODE_PATH` resolution order to favor project root, then
  archetypes, then existing environment.
  [#47](https://github.com/FormidableLabs/builder/issues/47)

## 2.1.3

* Just use `require()` for archetype package.json loading.
  [#32](https://github.com/FormidableLabs/builder/issues/32)

## 2.1.2

* Allow archetype discovery from siblings for any npm version.
  [#30](https://github.com/FormidableLabs/builder/issues/30)

## 2.1.1

* Fix bug with archetype discovery when npm-installed and npm v3.
  [#25](https://github.com/FormidableLabs/builder/issues/25)

## 2.1.0

* Support new `ARCHETYPE` + `ARCHETYPE-dev` architecture for better NPM 2 + 3

lib/task.js
"use strict";

/* eslint max-statements:[2, 30]*/

const path = require("path");
const _ = require("lodash");
const async = require("async");
const chalk = require("chalk");
const args = require("./args");
const Environment = require("./environment");
const log = require("./log");
const clone = require("./utils/clone");
const jsonParse = require("./utils/json").parse;
const Tracker = require("./utils/tracker");
const runner = require("./utils/runner");

// Paths for inference of "is builder task?"
const BUILDER_NODE_MODULES = path.normalize("node_modules/builder/bin/builder.js");
const BUILDER_BIN = path.normalize("node_modules/.bin/builder");

/**
 * Task wrapper.
 *
 * @param {Object} opts         Options object
 * @param {Object} opts.config  Configuration object
 * @param {Object} opts.env     Environment object to mutate (Default `Environment`)
 * @param {Array}  opts.argv    Arguments array (Default: `process.argv`)
 * @param {Object} opts.runner  Runner object.
 * @returns {void}
 */
const Task = module.exports = function (opts) {
  opts = opts || {};

  // Options.
  this._config = opts.config;
  this._env = opts.env || new Environment();

  // Infer parts.
  this.argv = opts.argv || process.argv;
  const parsed = args.general(this.argv);
  const remain = parsed.argv.remain;
  this._script = this.argv[1];
  this._action = remain[0];
  this._command = remain[1];
  this._commands = remain.slice(1);

  // State
  this._tracker = new Tracker();
  this._setupKilled = false;
  this._errors = [];

  // Validation.
  if (!this._config) {
    throw new Error("Configuration object required");
  }

  // Special flags that short circuit.
  if (parsed.help === true) {
    this._action = "help";
  } else if (parsed.version === true) {
    this._action = "version";
  } else if (remain.length === 0) {
    this._action = "help";
  }

  // Infer action.
  if (!_.includes(this.ACTIONS.concat(["version"]), this._action)) {
    throw new Error(`Invalid action: ${this._action
    } - Valid actions: ${this.ACTIONS.join(", ")}`);
  }
};

Task.prototype.ACTIONS = ["run", "concurrent", "envs", "help"];

Task.prototype.toString = function () {
  let cmd = this._command;
  if (this._action === "concurrent") {
    cmd = this._commands.join(", ");
  } else if (_.includes(["help", "version"], this._action)) {
    cmd = null;
  }

  return this._action + (cmd ? ` ${cmd}` : "");
};

/**
 * Is this task another builder command?
 *
 * Checks either `builder` or `node PATH/TO/BUILDER`.
 *
 * _Note_: Naive. Only captures `builder` at start of command.
 * See: https://github.com/FormidableLabs/builder/issues/93
 *
 * @param   {String} task   Task
 * @returns {Boolean}       Is this task a passthrough?
 */
Task.prototype.isBuilderTask = function (task) {
  // mac: '/PATH/TO/node_modules/.bin/builder'
  // win: 'X:\\PATH\\TO\\node_modules\\builder\\bin\\builder.js',
  const builder = path.basename(this._script, ".js");
  const taskParts = (task || "").trim().split(/\s+/);
  let taskBin = taskParts[0] || "";

  // Case: `builder run foo`
  // Compare directly if not a `node` script execution.
  if (taskBin.toLowerCase() !== "node") {
    return taskBin === builder;
  }

  // We know we're now in `node PATH/TO/BUILDER`.
  taskBin = path.resolve(taskParts[1] || "");

  // Case: `node CURRENT_BUILDER_PATH`
  return taskBin === path.resolve(this._script)

    // Case: `node node_modules/builder/bin/builder.js`
    || taskBin.substr(taskBin.length - BUILDER_NODE_MODULES.length) === BUILDER_NODE_MODULES

    // Case: `node node_modules/.bin/builder`
    || taskBin.substr(taskBin.length - BUILDER_BIN.length) === BUILDER_BIN;
};

/**
 * Is this task a simple passthrough to another builder command?
 *
 * @param   {String} task   Task
 * @returns {Boolean}       Is this task a passthrough?
 */
Task.prototype.isPassthrough = function (task) {
  const taskParts = task.split(/\s+/);
  const taskAction = taskParts[1];
  const taskCommand = taskParts[2]; // eslint-disable-line no-magic-numbers

  return this.isBuilderTask(task)
    && this._action === taskAction
    && this._command === taskCommand;
};

Task.prototype._findCommand = function (cmd) {
  const self = this;

  // Select first non-passthrough command.
  return _.find(this._config.getCommands(cmd), (obj) => !self.isPassthrough(obj.cmd));
};

/**
 * Get execution parameters, options, etc.
 *
 * @param   {String} cmd    Script command
 * @param   {String} action Action name (default `general`)
 * @returns {Object}        Commands obj `{ main: { archetype, cmd, opts }, pre, post }`
 */
Task.prototype.getRunParams = function (cmd, action) {
  // For external use:
  cmd = cmd || this._command;
  action = action || "general";

  // Require a `main` task.
  const main = this._findCommand(cmd);
  if (!main) {
    if (_.isUndefined(cmd)) {
      throw new Error("No task specified");
    }

    throw new Error(`Unable to find task for: ${cmd}`);
  }

  // Gather pre|post tasks, if they aren't already pre|post-prefixed.
  // **Note**: This follows `npm` behavior. `yarn` would run `prepre<task>`
  // when executing `yarn run pre<task>`.
  let pre = null;
  let post = null;
  if (!(/^(pre|post)/).test(cmd)) {
    pre = this._findCommand(`pre${cmd}`) || null;
    post = this._findCommand(`post${cmd}`) || null;
  }

  // Infer execution options and enhance objects.
  const taskArgs = args[action](this.argv);
  if (pre) {
    _.extend(pre, { opts: this.getPrePostOpts(pre, taskArgs, action) });
  }
  _.extend(main, { opts: this.getMainOpts(main, taskArgs) });
  if (post) {
    _.extend(post, { opts: this.getPrePostOpts(post, taskArgs, action) });
  }

  return {
    pre,
    main,
    post
  };
};

/**
 * Merge base options with custom options.
 *
 * @param   {Object} task   Task object `{ archetype, cmd }`
 * @param   {Object} opts   Custom options
 * @returns {Object}        Combined options
 */
Task.prototype.getMainOpts = function (task, opts) {
  return _.extend({
    _isBuilderTask: this.isBuilderTask(task.cmd),
    _archetypeName: task.archetypeName,
    _archetypePath: task.archetypePath
  }, opts);
};

/**
 * Merge base options with custom options and filter out non-pre|post things.
 *
 * @param   {Object} task   Task object `{ archetype, cmd }`
 * @param   {Object} opts   Custom options
 * @param   {String} action Action type to take.
 * @returns {Object}        Combined options
 */
Task.prototype.getPrePostOpts = function (task, opts, action) {
  if (!task) { return null; }

  // Action specific overrides.
  const overrides = {
    envs: {
      buffer: false
    }
  };

  // Generally applicable options.
  return _.extend({}, this.getMainOpts(task, opts), {
    tries: 1,
    setup: false,
    _customFlags: null
  }, overrides[action]);
};

// Expose require because with mock-fs and node 0.10, 0.12, the `require` itself
// is mocked.
Task.prototype._lazyRequire = require; // eslint-disable-line global-require

/**
 *
 * @param {String}    taskName  Setup task name
 * @param {Object}    shOpts    Shell options
 * @param {Function}  callback  Function `(err)` called on process end.
 * @returns {Object}            Process object or `null` if no setup.
 */
Task.prototype.setup = function (taskName, shOpts, callback) {
  // Lazy require because `task` is a dependency.
  const setup = this._lazyRequire("./utils/setup");

  // Short-circuit empty task.
  if (!taskName) { return null; }

  // Add, invoke, and hook to final callback if setup dies early.
  const self = this;
  const proc = this._tracker.add(setup.create(taskName, shOpts));
  if (proc) {
    // If setup exit happens before normal termination, kill everything.
    proc.on("exit", (code) => {
      self._setupKilled = true;
      callback(new Error(`Setup exited with code: ${code}`));
    });
  }

  return proc;
};

Task.prototype.trackAndRun = function () {
  return this._tracker.add(runner.run.apply(null, arguments));
};


// Wrap all callbacks to noop if the tracker or setup are already killed.
Task.prototype._skipIfKilled = function (fn) {
  const self = this;
  return function (cb) {
    return self._tracker.killed || self._setupKilled ? cb() : fn(cb);
  };
};

// Convenience wrapper for processing errors with `opts.bail`.
Task.prototype._errOrBail = function (opts, callback) {
  const self = this;
  return function (err) {
    if (err) {
      self._errors.push(err);
    }
    callback(opts.bail ? err : null); // Decide if failure ends the run.
  };
};

/**
 * Help.
 *
 * ```sh
 * $ builder help <action>
 * $ builder help <archetype1> <archetype2>
 * $ builder --help
 * $ builder -h
 * ```
 *
 * @param   {Function} callback   Callback function `(err)`
 * @returns {void}
 */
Task.prototype.help = function (callback) {
  // Arguments after `help` are action OR archetypes.
  const cmd = this._command;
  const flagsDisplay = chalk.green.bold("Flags");

  // One matching command is an action: `builder help run`
  const action = _.includes(this.ACTIONS, cmd) ? cmd : null;
  const actionDisplay = action ? chalk.red(action) : "<action>";
  const actions = this.ACTIONS.map((val) => val === action ? chalk.red(action) : val).join(", ");
  const actionFlags = action
    ? `\n\n${flagsDisplay}: ${actionDisplay}\n\n  ${args.help(cmd)}`
    : "";

  // No matched action means all string are archetypes: `builder help <arch1> <arch2>`
  const archetypes = action ? null : this._commands;

  // Display task configs if we have _some_.
  const taskConfigs = this._config.displayConfigs(archetypes);

  // Use raw stdout instead of logger so we can disable logger on "help".
  process.stdout.write(
    `${chalk.green.bold("Usage")}: \n\n  builder ${actionDisplay} <task(s)>`
    + `\n\n${chalk.green.bold("Actions")}: \n\n  ${actions
    }\n\n${flagsDisplay}: General\n\n  ${args.help()
    }${actionFlags
    }${taskConfigs ? `\n\n${chalk.green.bold("Task Configs")}: \n${taskConfigs}` : ""
    }\n\n${chalk.green.bold("Tasks")}: \n${this._config.displayScripts(archetypes)}\n`);

  callback();
};

/**
 * Version.
 *
 * ```sh
 * $ builder --version
 * $ builder -v
 * ```
 *
 * @param   {Function} callback   Callback function `(err)`
 * @returns {void}
 */
Task.prototype.version = function (callback) {
  process.stdout.write(`${args.version()}\n`);
  callback();
};

/**
 * Run.
 *
 * @param   {Function} callback   Callback function `(err)`
 * @returns {void}
 */
Task.prototype.run = function (callback) {
  // `builder run` -> `builder help`
  if (!this._command) {
    return void this.help(callback);
  }

  // Ensure only one call, so `--setup` process can call early for error.
  callback = _.once(callback);

  // Aliases
  const self = this;
  const shOpts = { env: this._env.env };

  // Get execution parameters.
  const params = this.getRunParams(this._command, "run");
  log.info(this._action, this._command + chalk.gray(` - ${params.main.cmd}`));

  // Task execution sequence
  async.series([
    // Execute `pre<task>`
    params.pre && self.trackAndRun.bind(self, params.pre.cmd, shOpts, params.pre.opts),

    // Execute `--setup=<setup-task>` (spawned for lifetime of <task>)
    function (cb) {
      self.setup(params.main.opts.setup, shOpts, callback);
      cb();
    },

    // Execute `<task>`
    function (cb) {
      const opts = _.extend({ tracker: self._tracker }, params.main.opts);

      return opts.tries === 1
        ? self.trackAndRun(params.main.cmd, shOpts, opts, cb)
        : runner.retry(params.main.cmd, shOpts, opts, cb);
    },

    // Execute `post<task>`
    params.post && self.trackAndRun.bind(self, params.post.cmd, shOpts, params.post.opts)
  ].filter(Boolean).map(self._skipIfKilled.bind(self)), callback);
};

/**
 * Concurrent.
 *
 * @param   {Function} callback   Callback function `(err)`
 * @returns {void}
 */
Task.prototype.concurrent = function (callback) {
  // Ensure only one call, so `--setup` process can call early for error.
  callback = _.once(callback);

  // Aliases
  const self = this;
  const shOpts = { env: this._env.env };

  // Get execution parameters (list).
  const cmds = this._commands;
  const paramsList = cmds.map((cmd) => self.getRunParams(cmd, "concurrent"));
  const firstParams = paramsList[0];
  if (!firstParams) {
    log.warn(this._action, "No tasks found");
    return void callback();
  }

  const paramsText = paramsList
    .map((t, i) => `\n * ${cmds[i]}${chalk.gray(` - ${t.main.cmd}`)}`)
    .join("");
  log.info(this._action, cmds.join(", ") + paramsText);

  // Use first task options to slice off general concurrent information.
  const queue = firstParams.main.opts.queue;

  // Start setup on first of **any** main tasks.
  const startSetup = _.once((opts) => opts.setup ? self.setup(opts.setup, shOpts, callback) : null);

  log.debug("concurrent", `Starting with queue size: ${chalk.magenta(queue || "unlimited")}`);
  async.mapLimit(paramsList, queue || Infinity, (params, concCb) => {
    // Task execution sequence
    async.series([
      // Execute `pre<task>`
      params.pre && self.trackAndRun.bind(self, params.pre.cmd, shOpts, params.pre.opts),

      // (ONLY FIRST ONE) Execute `--setup=<setup-task>` (spawned for lifetime of <task>)
      function (cb) {
        startSetup(params.main.opts, shOpts, callback);
        cb();
      },

      // Execute `<task1>`, `<task2>`, `...`
      function (cb) {
        const opts = _.extend({ tracker: self._tracker }, params.main.opts);

        return opts.tries === 1
          ? self.trackAndRun(params.main.cmd, shOpts, opts, cb)
          : runner.retry(params.main.cmd, shOpts, opts, cb);
      },

      // Execute `post<task>`
      params.post && self.trackAndRun.bind(self, params.post.cmd, shOpts, params.post.opts)

      // Apply --bail to **all** of this sequence by wrapping .series final callback.
    ].filter(Boolean).map(self._skipIfKilled.bind(self)),
    self._errOrBail(params.main.opts, concCb));
  }, callback);
};

// Return validated environment object from CLI or file.
Task.prototype._getEnvsList = function (parseObj) {
  // Parse envs.
  let envsList;
  let err;
  try {
    envsList = jsonParse(parseObj);
  } catch (parseErr) {
    err = parseErr;
  }

  // Validation
  if (!err && _.isEmpty(envsList)) {
    err = new Error("Empty/null JSON environments array.");
  } else if (!err && !_.isArray(envsList)) {
    err = new Error(`Non-array JSON environments object: ${JSON.stringify(envsList)}`);
  }

  if (err) {
    log.error(`${this._action}:json-error`,
      `Failed to load environments string / path with error: ${err}`);
    throw err;
  }

  return envsList;
};

/**
 * Run multiple environments.
 *
 * @param   {Function} callback   Callback function `(err)`
 * @returns {void}
 */
Task.prototype.envs = function (callback) {
  // Ensure only one call, so `--setup` process can call early for error.
  callback = _.once(callback);

  // Aliases
  const self = this;
  const shOpts = { env: this._env.env };

  // Get execution parameters.
  const params = this.getRunParams(this._command, "envs");
  log.info(this._action, this._command + chalk.gray(` - ${params.main.cmd}`));

  // Get list of environment variable objects.
  let envsList;
  try {
    envsList = this._getEnvsList({
      str: this._commands[1], // from CLI arguments
      path: params.main.opts.envsPath // from `--envs-path`
    });
  } catch (err) {
    return void callback(err);
  }

  // Task execution sequence
  async.series([
    // Execute `pre<task>`
    params.pre && self.trackAndRun.bind(self, params.pre.cmd, shOpts, params.pre.opts),

    // Execute `--setup=<setup-task>` (spawned for lifetime of <task>)
    function (cb) {
      self.setup(params.main.opts.setup, shOpts, callback);
      cb();
    },

    // Execute `<task>` w/ each env var.
    function (cb) {
      const cmd = params.main.cmd;
      const cmdStr = runner.cmdStr;
      const opts = _.extend({ tracker: self._tracker }, params.main.opts);
      const taskEnvs = clone(envsList);
      const queue = opts.queue;

      log.debug("envs", `Starting with queue size: ${chalk.magenta(queue || "unlimited")}`);
      async.mapLimit(taskEnvs, queue || Infinity, (taskEnv, envCb) => {
        // Add each specific set of environment variables.
        // Clone `shOpts` to turn `env` into a plain object: in Node 4+
        // `process.env` is a special object which changes merge behavior.
        const taskShOpts = _.merge(clone(shOpts), { env: taskEnv });
        const taskOpts = _.extend({ taskEnv }, opts);

        log.info("envs", `Starting ${cmdStr(cmd, taskOpts)}`);
        runner.retry(cmd, taskShOpts, taskOpts, self._errOrBail(taskOpts, envCb));
      }, cb);
    },

    // Execute `post<task>`
    params.post && self.trackAndRun.bind(self, params.post.cmd, shOpts, params.post.opts)
  ].filter(Boolean).map(self._skipIfKilled.bind(self)), callback);
};

/**
 * Clean up task state.
 *
 * - Terminate all existing processes.
 * - Drain and log all accumulated errors.
 *
 * @param {Function}  callback  Callback `(err)`
 * @returns {Function}          Wrapped callback
 */
Task.prototype.finish = function (callback) {
  const errors = this._errors;

  this._tracker.kill(() => {
    const errsText = errors
      .map((errItem) => `  * ${chalk.gray(errItem.name)}: ${chalk.red(errItem.message)}`)
      .join("\n");
    if (errors.length > 1) {
      log.error("finish", `Hit ${chalk.red(errors.length)} errors: \n${errsText}`);
    }

    callback(errors[0]);
  });
};

/**
 * Execute action and cleanup.
 *
 * **Note**: All tasks (e.g., `run`, `concurrent`) through this gate.
 *
 * @param   {Function} callback   Callback function `(err)`
 * @returns {Object}              Process object for `run` or `null` otherwise / for errors
 */
Task.prototype.execute = function (callback) {
  const self = this;
  const action = this._action;

  // Check task action method exists.
  if (!this[action]) {
    return void callback(new Error(`Unrecognized action: ${action}`));
  }

  // Call action.
  this[action]((err) => {
    // Add errors, if any.
    if (err) {
      self._errors.push(err);
    }

    // Unconditionally cleanup.
    self.finish(callback);
  });
};
